src/jit/lclvars.cpp

```diff
@@ -1542,17 +1535,59 @@ void Compiler::lvaCanPromoteStructType(CORINFO_CLASS_HANDLE    typeHnd,
                 var_types simdBaseType = getBaseTypeAndSizeOfSIMDType(pFieldInfo->fldTypeHnd, &simdSize);
                 if (simdBaseType != TYP_UNKNOWN)
                 {
-                    varType             = getSIMDTypeForSize(simdSize);
-                    pFieldInfo->fldType = varType;
+                    pFieldInfo->fldType = getSIMDTypeForSize(simdSize);
                     pFieldInfo->fldSize = simdSize;
                 }
             }
 #endif // FEATURE_SIMD
 
             if (pFieldInfo->fldSize == 0)
             {
-                // Non-primitive struct field. Don't promote.
-                return;
+                // Non-primitive struct field.
```

Comment:
I think not because there other TYP_* whose genTypeSize() is zero as per typelist.h, which are not necessarily structs.

I would add check that for == TYP_STRUCT in addition because the new logic that i have added is assuming that fldSize == 0 means TYP_STRUCT.